chain/chain/src/state_sync.rs

```diff
@@ -0,0 +1,191 @@
+use near_chain_primitives::error::Error;
+use near_primitives::hash::CryptoHash;
+use near_primitives::types::EpochId;
+use near_store::{DBCol, Store, StoreUpdate};
+
+use borsh::BorshDeserialize;
+use std::collections::BinaryHeap;
+
+use crate::types::BlockHeader;
+use crate::ChainStoreAccess;
+
+// Used for storing block headers in a heap that returns the lowest heights first
+#[derive(PartialEq, Eq)]
+struct BlockHeaderByHeight<'a>(&'a BlockHeader);
+
+impl<'a> PartialOrd for BlockHeaderByHeight<'a> {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl<'a> Ord for BlockHeaderByHeight<'a> {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        other.0.height().cmp(&self.0.height())
+    }
+}
+
+fn get_epoch_new_chunks(
+    store: &Store,
+    block_hash: &CryptoHash,
+) -> Result<Option<Vec<usize>>, Error> {
+    Ok(store.get_ser(DBCol::StateSyncNewChunks, block_hash.as_ref())?)
+}
+
+fn iter_epoch_new_chunks_keys<'a>(
+    store: &'a Store,
+) -> impl Iterator<Item = Result<CryptoHash, std::io::Error>> + 'a {
+    store
+        .iter(DBCol::StateSyncNewChunks)
+        .map(|item| item.and_then(|(k, _v)| CryptoHash::try_from_slice(&k)))
+}
+
+fn iter_state_sync_hashes_keys<'a>(
+    store: &'a Store,
+) -> impl Iterator<Item = Result<EpochId, std::io::Error>> + 'a {
+    store
+        .iter(DBCol::StateSyncHashes)
+        .map(|item| item.and_then(|(k, _v)| EpochId::try_from_slice(&k)))
+}
+
+fn save_epoch_new_chunks<T: ChainStoreAccess>(
+    chain_store: &T,
+    store_update: &mut StoreUpdate,
+    header: &BlockHeader,
+) -> Result<(), Error> {
+    let Some(mut num_new_chunks) = get_epoch_new_chunks(chain_store.store(), header.prev_hash())?
+    else {
+        // This might happen in the case of epoch sync where we save individual headers without having all
+        // headers that belong to the epoch.
```

Comment:
It will actually, if you add a print statement here and run `test_epoch_sync_with_expired_transactions` with `--features nightly` you'll see it triggered after adding these headers: https://github.com/near/nearcore/blob/a86de63de736e2468f16fce333b5e5c88e7b4664/chain/client/src/sync/epoch.rs#L697-L704 

Looks like we add some individual headers before the normal header sync starts, but then after the normal header sync starts we have all the headers like you say, and then this all works